packages/react-tabs/Spec.md

```diff
@@ -63,211 +72,270 @@ The existing components are:
+The size affects the padding and spacing between elements.
+Size can also affect the default icon (and badge) size.
+The default size is medium.
 
 ```tsx
-<TabList>{/* ... */}</TabList>
+<TabList size="small">
+  {/* ... */}
+</TabList>
 
-<TabList appearance="transparent">{/* ... */}</TabList>
+<TabList size="medium">
+  {/* ... */}
+</TabList>
 ```
 
+### With Icon
+
+When the icon slot is filled, an icon is positioned before the tab content.
+
 ```tsx
-<TabList appearance="subtle">{/* ... */}</TabList>
+<TabList>
+  <Tab icon={<CheckboxComposite />} value="allowed">
+    Allowed
+  </Tab>
+  <Tab icon={<BlockedSite />} value="blocked">
+    Blocked
+  </Tab>
+</TabList>
 ```
 
-### Badges
+An icon only tab is displayed when the icon slot is filled and tab content is omitted.
+
+```tsx
+<TabList>
+  <Tab icon={<CheckboxComposite />} value="allowed" />
+  <Tab icon={<BlockedSite />} value="blocked" />
+</TabList>
+```
 
-> TODO: How to support multiple badge types as a slot?
-> TODO: How to specify icon for basic badge in a slot?
+### With Badge
+
+Adding a badge to a tab is done through the BadgeTab variant.
+This follows a extensible approach of MenuItem rather than defining a badge slot on the Tab component.
+The BadgeTab has a badge slot (BadgeProps) and leverages hooks from Badge.
+
+The badge is positioned relative to the icon depending on the verticalTabContent and size props.
+The BadgeTab variant support content and icon in addition to the badge.
 
 ```tsx
 <TabList>
-  <Tab>Files</Tab>
-  <Tab>Search</Tab>
-  <Tab badge={<CounterBadge count={5} />}>Messages</Tab>
+  <BadgeTab badge="24">Inbox</BadgeTab>
+  <BadgeTab badge="24" icon={<MailMessage />}>
+    Inbox
+  </BadgeTab>
 </TabList>
 ```
 
-### Overflow
+### TabPanel (future)
+
+The initial vNext package will **not** provide a tab panel, but the tab list will be designed to support a tab panel in the future through composition.
+
+Why not now?
 
-Handling overflow is a built-in feature of tabs and replaceable via a slot.
+- There are many approaches to handling tab panel content: show/hide, conditional render, navigation through a router
+- We're not providing a manilla folder tab appearance, so there is no need for the tab panel to support styling for when the components are co-located.
+- Creating a specific tab panel is trivial, creating a general-purpose tab panel is much more involved.
+
+The tab panels should be able to be placed independently of the TabList.
+Here's a possible approach:
 
 ```tsx
-<TabList overflow={</MyOverflowButton}>
-  <Tab>One</Tab>
-  <Tab>Two</Tab>
-  <Tab>Three</Tab>
+<TabList onTabSelected={(e,data) => setCurrentTab(data.value)}>
+  <Tab value="tab1">One</Tab>
+  <Tab value="tab2">Two</Tab>
+  <Tab value="tab3">Three</Tab>
 </TabList>
+
+<TabPanels value={currentTab}>
+  <TabPanel value="tab1">Tab Panel #1</TabPanel>
+  <TabPanel value="tab2">Tab Panel #2</TabPanel>
+  <TabPanel value="tab3">Tab Panel #3</TabPanel>
+</TabPanels>
```

Comment:
Another option could be to have controlling component like in `Menu` case:

```tsx
<Tabs>
  <TabList>
    <Tab value="tab1">One</Tab>
  </TabList>

  <TabPanels>
    <TabPanel value="tab1">Tab Panel #1</TabPanel>
  </TabPanels>
</Tabs>
```

In this case `Tabs` controls state and we can implement autocontrolled pattern.

---

BTW, in Northstar for naming we used `Group` instead of `s`: `TabPanels` => `TabPanelGroup`.

Reply:
I like this idea and was thinking the same thing initially.  I got some feedback that it might not be the right approach because MenuTrigger and MenuPopover are independent. However, it seems like the same case of a wrapper needing to coordinate.  I'll update the spec to go with this approach.